File: DSA/src/Questions/DP/BuySellStocks/WithCoolDown.java
package Questions.DP.BuySellStocks;

import java.util.Arrays;

public class WithCoolDown {

    // Recursive function to calculate the maximum profit
    static int getAns(int[] Arr, int ind, int buy, int n, int[][] dp) {
        // Base case
        if (ind >= n) {
            return 0;
        }

        // If the result is already calculated, return it
        if (dp[ind][buy] != -1) {
            return dp[ind][buy];
        }

        int profit = 0;

        if (buy == 0) { // We can buy the stock
            profit = Math.max(0 + getAns(Arr, ind + 1, 0, n, dp),
                    -Arr[ind] + getAns(Arr, ind + 1, 1, n, dp));
        }

        if (buy == 1) { // We can sell the stock
            profit = Math.max(0 + getAns(Arr, ind + 1, 1, n, dp),
                    Arr[ind] + getAns(Arr, ind + 2, 0, n, dp));
        }

        // Store the result in dp and return it
        dp[ind][buy] = profit;
        return profit;
    }

    static int stockProfitRecursive(int[] Arr) {
        int n = Arr.length;
        int[][] dp = new int[n][2];

        // Initialize dp array with -1 to mark states as not calculated yet
        for (int[] row : dp) {
            Arrays.fill(row, -1);
        }

        int ans = getAns(Arr, 0, 0, n, dp);
        return ans;
    }
    // Time Complexity: O(N*2)

    // Reason: There are N*2 states therefore at max ‘N*2’ new problems will be
    // solved and we are running a for loop for ‘N’ times to calculate the total sum

    // Space Complexity: O(N*2) + O(N)

    // Reason: We are using a recursion stack space(O(N)) and a 2D array ( O(N*2)).

    // Function to calculate the maximum profit from stock trading
    static int stockProfitTabulation(int[] Arr) {
        int n = Arr.length;
        // Each state is represented by two variables: ind and buy
        // ind: The index - 1 of the stock we are currently processing
        // buy: 0 if we can buy the stock, 1 if we can sell the stock
        // dp[ind + 1][buy] stores the maximum profit we can make at index 'ind' with
        int dp[][] = new int[n + 2][2];

        // Iterate through the array backwards
        for (int ind = n - 1; ind >= 0; ind--) {
            for (int buy = 0; buy <= 1; buy++) {
                int profit = 0;

                if (buy == 0) { // We can buy the stock
                    profit = Math.max(0 + dp[ind + 1][0], -Arr[ind] + dp[ind + 1][1]);
                }

                if (buy == 1) { // We can sell the stock
                    profit = Math.max(0 + dp[ind + 1][1], Arr[ind] + dp[ind + 2][0]);
                }

                dp[ind][buy] = profit;
            }
        }

        // The maximum profit is stored in dp[0][0]
        return dp[0][0];
    }
    // Time Complexity: O(N*2)

    // Reason: There are two nested loops that account for O(N*2) complexity.

    // Space Complexity: O(N*2)

    // Reason: We are using an external array of size ‘N*2’. Stack Space is
    // eliminated.

    // Function to calculate the maximum profit from stock trading
    static int stockProfit(int[] Arr) {
        int n = Arr.length;
        int[] cur = new int[2];
        int[] front1 = new int[2];
        int[] front2 = new int[2];

        // Iterate through the array backwards
        for (int ind = n - 1; ind >= 0; ind--) {
            for (int buy = 0; buy <= 1; buy++) {
                int profit = 0;

                if (buy == 0) { // We can buy the stock
                    profit = Math.max(0 + front1[0], -Arr[ind] + front1[1]);
                }

                if (buy == 1) { // We can sell the stock
                    profit = Math.max(0 + front1[1], Arr[ind] + front2[0]);
                }

                cur[buy] = profit;
            }

            // Update front1 and front2 arrays
            System.arraycopy(front1, 0, front2, 0, 2);
            System.arraycopy(cur, 0, front1, 0, 2);
        }

        // The maximum profit is stored in cur[0]
        return cur[0];
    }
    // Time Complexity: O(N*2)

    // Reason: There are two nested loops that account for O(N*2) complexity

    // Space Complexity: O(1)

    // Reason: We are using three external arrays of size ‘2’.

    public static void main(String[] args) {
        int[] Arr = { 1, 2, 3, 0, 2 };
        System.out.println(stockProfitRecursive(Arr)); // 3
        System.out.println(stockProfitTabulation(Arr)); // 3
        System.out.println(stockProfit(Arr)); // 3
    }

}

File: DSA/Python/Questions/DP/BuySellStocks/with_cool_down.py
"""
# Question: Best Time to Buy and Sell Stock with Cooldown
# Link: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-cooldown/

# Problem Statement:
# You are given an array prices where prices[i] is the price of a given stock on the ith day.
# After selling stock, you cannot buy stock on the next day (i.e., cooldown one day).
# Find the maximum profit you can achieve.

# Example:
# Input: prices = [1,2,3,0,2]
# Output: 3
# Explanation: Buy on day 1 (price = 1) and sell on day 2 (price = 2), profit = 1.
# Then buy on day 4 (price = 0) and sell on day 5 (price = 2), profit = 2.
# Total profit = 3.
"""

from typing import List

# 1. Recursive Solution with Memoization
"""
Algorithm:
1. For each day, three possible states:
   - Buy stock if not in cooldown
   - Sell stock and enter cooldown
   - Skip current day
2. Use memoization to store results
3. Handle cooldown by skipping next day after sell

Time Complexity: O(n*2)
Space Complexity: O(n*2) + O(n) recursion stack
"""
def max_profit_recursive(prices: List[int]) -> int:
    n = len(prices)
    dp = {}
    
    def solve(index: int, can_buy: bool) -> int:
        if index >= n:
            return 0
            
        if (index, can_buy) in dp:
            return dp[(index, can_buy)]
            
        if can_buy:
            profit = max(-prices[index] + solve(index + 1, False),
                        solve(index + 1, True))
        else:
            profit = max(prices[index] + solve(index + 2, True),
                        solve(index + 1, False))
                        
        dp[(index, can_buy)] = profit
        return profit
        
    return solve(0, True)

# 2. Tabulation Solution
"""
Algorithm:
1. Create 2D DP array [n+2][2]
2. Fill table bottom-up
3. For each state:
   - Buy: max(-price[i] + dp[i+1][0], dp[i+1][1])
   - Sell: max(price[i] + dp[i+2][1], dp[i+1][0])
4. Handle cooldown in sell state

Time Complexity: O(n*2)
Space Complexity: O(n*2)
"""
def max_profit_tabulation(prices: List[int]) -> int:
    n = len(prices)
    dp = [[0] * 2 for _ in range(n + 2)]
    
    for i in range(n-1, -1, -1):
        for buy in range(2):
            if buy:
                dp[i][buy] = max(-prices[i] + dp[i+1][0],
                                dp[i+1][1])
            else:
                dp[i][buy] = max(prices[i] + dp[i+2][1],
                                dp[i+1][0])
                                
    return dp[0][1]

# 3. Space Optimized Solution
"""
Algorithm:
1. Use three arrays for current and next two days
2. Process each day updating arrays
3. Handle cooldown period using two ahead states

Time Complexity: O(n*2)
Space Complexity: O(1)
"""
def max_profit_optimized(prices: List[int]) -> int:
    n = len(prices)
    front1 = [0] * 2
    front2 = [0] * 2
    curr = [0] * 2
    
    for i in range(n-1, -1, -1):
        for buy in range(2):
            if buy:
                curr[buy] = max(-prices[i] + front1[0],
                               front1[1])
            else:
                curr[buy] = max(prices[i] + front2[1],
                               front1[0])
                               
        front2 = front1[:]
        front1 = curr[:]
        
    return curr[1]

def main():
    test_cases = [
        [1, 2, 3, 0, 2],              # Regular case
        [1],                          # Single element
        [1, 2],                       # Two elements
        [1, 2, 3, 4, 5],              # Strictly increasing
        [5, 4, 3, 2, 1],              # Strictly decreasing
        [2, 1, 4, 5, 2, 9, 7]         # Multiple peaks and valleys
    ]
    
    for i, prices in enumerate(test_cases):
        print(f"\nTest Case {i + 1}: {prices}")
        print(f"Recursive Solution: {max_profit_recursive(prices)}")
        print(f"Tabulation Solution: {max_profit_tabulation(prices)}")
        print(f"Optimized Solution: {max_profit_optimized(prices)}")

if __name__ == "__main__":
    main()
